In [1]:
import numpy as np
import pandas_datareader as pdr
import datetime as dt
import pandas as pd

Portfolios

In [2]:
import pandas_datareader.data as pdr
import yfinance as yf
yf.pdr_override()
from datetime import datetime
data = pdr.get_data_yahoo('^GSPC', datetime(1970, 1, 1))

[*********************100%***********************]  1 of 1 completed


In [3]:
tickers = ['AAPL', 'MSFT', 'AMZN', 'IBM']
start = dt.datetime(2020, 1, 1)

data = pdr.get_data_yahoo(tickers, start)

[*********************100%***********************]  4 of 4 completed


In [4]:
investment = 100_000

In [5]:
data.head()

Adj Close                                         Close  \
                 AAPL       AMZN         IBM        MSFT       AAPL   
Date                                                                  
2020-01-02  73.347939  94.900497  108.749641  155.422043  75.087502   
2020-01-03  72.634842  93.748497  107.882347  153.486786  74.357498   
2020-01-06  73.213623  95.143997  107.689606  153.883469  74.949997   
2020-01-07  72.869293  95.343002  107.761864  152.480438  74.597504   
2020-01-08  74.041489  94.598503  108.661308  154.909195  75.797501   

                                                    High             ...  \
                 AMZN         IBM        MSFT       AAPL       AMZN  ...   
Date                                                                 ...   
2020-01-02  94.900497  129.464630  160.619995  75.150002  94.900497  ...   
2020-01-03  93.748497  128.432129  158.619995  75.144997  94.309998  ...   
2020-01-06  95.143997  128.202682  159.029999  74.989998  95.184502  ...   
2020-01-07  95.343002  128.288712  157.580002  75.224998  95.694504  ...   
2020-01-08  94.598503  129.359467  160.089996  76.110001  95.550003  ...   

                   Low                   Open                         \
                   IBM        MSFT       AAPL       AMZN         IBM   
Date                                                                   
2020-01-02  128.843216  158.330002  74.059998  93.750000  129.063095   
2020-01-03  127.686424  158.059998  74.287498  93.224998  127.695984   
2020-01-06  127.342255  156.509995  73.447502  93.000000  127.552582   
2020-01-07  127.533463  157.320007  74.959999  95.224998  127.810707   
2020-01-08  128.030594  157.949997  74.290001  94.902000  128.594650   

                           Volume                               
                  MSFT       AAPL      AMZN      IBM      MSFT  
Date                                                            
2020-01-02  158.779999  135480400  80580000  3293436  22622100  
2020-01-03  158.320007  146322800  75288000  2482890  21116200  
2020-01-06  157.080002  118387200  81236000  2537073  20813700  
2020-01-07  159.320007  108872000  80898000  3232977  21634100  
2020-01-08  158.929993  132079200  70160000  4545916  27746500  

[5 rows x 24 columns]

In [6]:
data=data['Adj Close']

In [7]:
data.head()

AAPL       AMZN         IBM        MSFT
Date                                                    
2020-01-02  73.347939  94.900497  108.749641  155.422043
2020-01-03  72.634842  93.748497  107.882347  153.486786
2020-01-06  73.213623  95.143997  107.689606  153.883469
2020-01-07  72.869293  95.343002  107.761864  152.480438
2020-01-08  74.041489  94.598503  108.661308  154.909195

In [8]:
portfolios = [.25, .40, .20, .15]

In [9]:
np.sum(portfolios)

1.0

In [10]:
(data/data.iloc[0])*portfolios*investment

AAPL          AMZN           IBM          MSFT
Date                                                              
2020-01-02  25000.000000  40000.000000  20000.000000  15000.000000
2020-01-03  24756.947287  39514.438614  19840.497072  14813.225628
2020-01-06  24954.219746  40102.633711  19805.050262  14851.510038
2020-01-07  24836.857949  40186.513198  19818.339133  14716.101600
2020-01-08  25236.390457  39872.711174  19983.754773  14950.504328
...                  ...           ...           ...           ...
2023-05-31  60414.104177  50823.759659  23648.812936  31693.382287
2023-06-01  61382.091961  51746.829557  23875.022599  32097.762371
2023-06-02  61675.215608  52370.642244  24353.183411  32369.925245
2023-06-05  61208.264817  52813.212338  24393.643539  32422.042230
2023-06-06  61082.155234  53365.368583  24402.839533  32203.925508

[863 rows x 4 columns]

Monte Carlo Simulation

In [11]:
weight = np.random.random(4)
weight /= weight.sum()
weight

array([0.33512762, 0.37424419, 0.21777684, 0.07285135])

In [12]:
weight.sum()

0.9999999999999999

In [13]:
np.sum((data/data.iloc[0])*portfolios*investment, axis=1)

Date
2020-01-02    100000.000000
2020-01-03     98925.108601
2020-01-06     99713.413757
2020-01-07     99557.811881
2020-01-08    100043.360732
                  ...      
2023-05-31    166580.059059
2023-06-01    169101.706488
2023-06-02    170768.966508
2023-06-05    170837.162924
2023-06-06    171054.288858
Length: 863, dtype: float64

In [14]:
np.sum((data/data.iloc[0])*weight*investment, axis=1)

Date
2020-01-02    100000.000000
2020-01-03     98955.496985
2020-01-06     99750.260617
2020-01-07     99620.119883
2020-01-08    100156.063115
                  ...      
2023-05-31    169680.505415
2023-06-01    172284.450753
2023-06-02    173913.876393
2023-06-05    173771.365518
2023-06-06    174022.996856
Length: 863, dtype: float64

In 3 years time, an investment of 100,000 Euro in a portfolio made of 25% Apple, 40% Amazon, 20% IBM and 15% Microsoft
gave us an actual return of 171,054.29 Euro while the Monte Carlo Simulation gives us a porfolio value of 158,923.10